mllib/src/main/scala/org/apache/spark/ml/feature/MaxAbsScaler.scala

```diff
@@ -70,14 +67,40 @@ class MaxAbsScaler @Since("2.0.0") (@Since("2.0.0") override val uid: String)
   @Since("2.0.0")
   override def fit(dataset: Dataset[_]): MaxAbsScalerModel = {
     transformSchema(dataset.schema, logging = true)
-    val input: RDD[OldVector] = dataset.select($(inputCol)).rdd.map {
-      case Row(v: Vector) => OldVectors.fromML(v)
-    }
-    val summary = Statistics.colStats(input)
-    val minVals = summary.min.toArray
-    val maxVals = summary.max.toArray
-    val n = minVals.length
-    val maxAbs = Array.tabulate(n) { i => math.max(math.abs(minVals(i)), math.abs(maxVals(i))) }
+
+    val maxAbs = dataset.select($(inputCol)).rdd.map {
+      row => row.getAs[Vector](0)
+    }.treeAggregate[Array[Double]](Array.emptyDoubleArray)(
+      seqOp = {
+        case (max, vec) if max.isEmpty =>
+          vec.toArray.map(math.abs)
+        case (max, vec) =>
+          require(max.length == vec.size,
+            s"Dimensions mismatch when adding new sample: ${max.length} != ${vec.size}")
+          vec.foreachActive {
+            case (i, v) if v != 0.0 =>
+              val av = math.abs(v)
+              if (av > max(i)) {
+                max(i) = av
+              }
+            case _ =>
+          }
+          max
+      }, combOp = {
```

Comment:
I think it may make the code clearer to move the seqOp and combOp to separate methods